### PR TITLE
WIP Record reader and high level read support

### PR DESCRIPTION
**Pull request by sadikovi**
PR adds read support, and it is for review. If everything is okay, I will open another clean PR to merge the changes.

Most of the API and code for reading records is in `src/record`.

> DO NOT MERGE YET

### CODE REVIEW

**Comment by sunchao** [comment]
Sorry for keep you waiting - I'm still reviewing this but it will take some time.

**Comment by sadikovi** [comment]
@sunchao thanks for taking a look at this PR! No worries, take as much time as you need. I am looking at something else meanwhile:).

File: src/bin/read.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+extern crate parquet;
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read.rs]: I actually prefer having them separate in `bin` directory for now: `dump-schema` is for inspecting schema and `read`/`cat` for listing records in a file. Then later we can move those files into a subproject `parquet-rs-tools`. Would you like this approach?

File: src/bin/read.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+use std::fs::File;
+use std::path::Path;
+use std::process;
+
+use parquet::file::reader::{FileReader, SerializedFileReader};
+use parquet::schema::printer::{print_file_metadata, print_schema};
+use parquet::schema::parser::parse_message_type;
+
+fn main() {
+  let args: Vec<_> = env::args().collect();
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read.rs]: I tried removing it, but then compiler complains that it cannot infer the type. I took it from `dump-schema` - it also has this line.

**Comment by sadikovi** [comment]
@sunchao As I am fixing your comments, would you comment on the overall direction of the API? 
My idea is that anyone who wants to use this API, would just map `Row` into whatever struct they need.

Feel free to leave more comments on this PR.

File: src/bin/read.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply src/bin/read.rs]: OK. Sounds fine. 

Another problem: there's name collision since read is already taken by the Shell built-in command, and if you install the binary via `cargo install`, you'll not be able to use this one.

File: src/record/triplet.rs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    match self.rep_levels {
+      Some(ref vec) => vec[self.curr_triplet_index],
+      None => self.max_rep_level
+    }
+  }
+
+  /// Consumes and advances to the next triplet.
+  /// Returns true, if there are more records to read, false there are no records left.
+  fn consume(&mut self) -> Result<bool> {
+    self.curr_triplet_index += 1;
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/record/triplet.rs]: I did not do it, because I wanted to trigger this method on the very first read - adding check would imply making it an option or something similar.

File: src/bin/read.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read.rs]: Aaa, okay. I will rename it to something that does not have conflicts, thanks!

File: src/bin/read.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read.rs]: Actually, I will try adding some code to parse command-line arguments or use a crate, since we would need that functionality.

**Comment by sunchao** [comment]
Thanks @sadikovi . I'll put more thought on the APIs. Currently I'm thinking:
1. maybe we need some APIs for the `Row` struct - something like `getField`, `getLength` etc, for different type of rows.
2. whether we should have a vectorized API that return batch of rows. 
3. whether the interface `Row` should be more general so future implementations such as Arrow can also share the same interface. One idea is that we ca replace the `Vec` occurrences in `Row` with some traits such as `Array`, or `Map`, which could be implementation specific. The Arrow integration requires a vectorized API though.

Let me know what you think. These do not need to be done in this PR though.

**Comment by sadikovi** [comment]
Sounds good. I am not very familiar with Arrow to make any compatible changes. We should be able to extend it in the future. 

Should I rename `Row` into `Record`?

**Comment by sadikovi** [comment]
Found that it still panics on some files, will investigate.

**Comment by sunchao** [comment]
> Found the problem. It is in column/reader.rs. We should account for buffering of a page when computing iter_batch_size. Will open PR for this.

Cool. Glad you found out the problem.

**Comment by sadikovi** [comment]
I still can't figure out how to get a global file iterator...

**Comment by sunchao** [comment]
@sadikovi : OK, let me try to modify the code and think about how to make it work. Basically the global file iterator should return an iterator of all rows in all the row groups, right?

BTW: is this still caused by the lifetime issue you mentioned before?

**Comment by sadikovi** [comment]
When I try adding iterator, I get different lifetime error, caused by row group reader.

Thanks! My plan is returning iterator of all rows in all row groups. Something tells me that the code in record/reader.rs needs to be changed significantly. I will also play with it this weekend.

**Comment by sadikovi** [comment]
Hi @sunchao. I fixed it! Now I can return global file iterator! This is really cool.

I had to remove lifetime parameter from column reader. It actually does not break anything, after this I can easily add file iterator and return it from a file reader.

Here is my code for file iterator:
```rust

pub struct FileRowIter<'a> {
  descr: SchemaDescPtr,
  file_reader: &'a FileReader,
  curr_row_group: usize,
  num_row_groups: usize,
  iter: Option<RowIter>
}

impl<'a> FileRowIter<'a> {
  pub fn new(descr: SchemaDescPtr, file_reader: &'a FileReader) -> Self {
    let num_row_groups = file_reader.num_row_groups();

    Self {
      descr: descr,
      file_reader: file_reader,
      curr_row_group: 0,
      num_row_groups: num_row_groups,
      iter: None
    }
  }
}

impl<'a> Iterator for FileRowIter<'a> {
  type Item = Row;

  fn next(&mut self) -> Option<Row> {
    let mut row = None;
    if let Some(ref mut iter) = self.iter {
      row = iter.next();
    }

    while row.is_none() && self.curr_row_group < self.num_row_groups {
      let row_group_reader = self.file_reader.get_row_group(self.curr_row_group).unwrap();
      self.curr_row_group += 1;
      let mut iter = row_group_reader.get_row_iter(self.descr.clone());
      row = iter.next();
      self.iter = Some(iter);
    }

    row
  }
}
```

I will push commits this weekend with final version.

Can you advise me if it is okay to remove lifetime from column reader (including impl)? Thanks.

**Comment by sunchao** [comment]
@sadikovi : glad you got it working! 

> Can you advise me if it is okay to remove lifetime from column reader (including impl)? Thanks.

Yes I think it's safe to remove the lifetime - it is not associated with anything right now. Could you submit a separate PR for this?

**Comment by sadikovi** [comment]
@sunchao I am thrilled to say that read API and the code have been working for me (and I assume for you) reasonably well. I rebased the branch to capture all recent changes, made some additional changes to the code and tested manually on test files, including large ones.

Would like to know your opinion on the current state. Feel free to make any comments, try `read-file`, etc. If you think it is ready to be added to the main repository, then I am going to close this PR and open a clean one with the same code and unit tests.

Also, could you give a suggestion on how I should test different types of records, e.g. structs, maps or lists, or lists of maps, etc. So far I have been manually testing it on test files. Should I add those test files to the repository? Or what is the best way of doing it? Thanks!

**Comment by sunchao** [comment]
@sadikovi great! I'm thrilled too as this will be a great feature! I think it is ready for the main repo, so feel free to close this WIP.

Testing is a little tricky as we don't have high-level APIs to generate the test records. I'm thinking maybe we can use the example Parquet files for now (e.g., write test cases reading from the example file and check result). We should also open a follow-up issue to add unit tests on this in future. What do you think?

As for the testing files, I think what you have is good. In addition, I also found that the Impala testing data for complex types to be useful: https://github.com/apache/impala/tree/master/testdata/ComplexTypesTbl. It covers a lot of edge cases.

**Comment by sadikovi** [comment]
Thanks @sunchao. Yes, let's use some subset of files to test for now. I will try adding unit tests wherever I can.

Tested on Impala test files that you provided with the link - and it works! Will start working on clean PR, thanks!
